Implement sidePanel setPanelBehavior/getPanelBehavior API methods.

AriYoung00 opened the pull request:
#### 7b5ac8389382fefbf45f9cb626343d7b94d60116
<pre>
Implement sidePanel setPanelBehavior/getPanelBehavior API methods.
<a href="https://webkit.org/b/280021">https://webkit.org/b/280021</a>
<a href="https://rdar.apple.com/135894275">rdar://135894275</a>

Reviewed by Timothy Hatcher.

This patch implements the sidePanel.setPanelBehavior and
sidePanel.getPanelBehavior API methods. Additionally, it adds two tests
to verify the functionality of these APIs.

* Source/WebKit/DerivedSources-input.xcfilelist: Add
  WebExtensionActionClickbehavior.serialization.in.
* Source/WebKit/DerivedSources.make: Add
  WebExtensionActionClickBehavior.serialization.in.
* Source/WebKit/Scripts/webkit/messages.py:
(types_that_cannot_be_forward_declared): Add
WebKit::WebExtensionActionClickBehavior to
types_that_cannot_be_forward_declared.
* Source/WebKit/Shared/Extensions/WebExtensionActionClickBehavior.h: Added.
* Source/WebKit/Shared/Extensions/WebExtensionActionClickBehavior.serialization.in: Added.
* Source/WebKit/UIProcess/Extensions/Cocoa/API/WebExtensionContextAPISidebarCocoa.mm:
(WebKit::WebExtensionContext::sidebarSetActionClickBehavior): Add
IPC message receiver.
(WebKit::WebExtensionContext::sidebarGetActionClickBehavior): Add IPC
message receiver.
* Source/WebKit/UIProcess/Extensions/Cocoa/WebExtensionContextCocoa.mm:
(WebKit::WebExtensionContext::performAction): Add case to open the
sidebar if configured and it is possible.
* Source/WebKit/UIProcess/Extensions/WebExtensionContext.h: Add
  prototypes for new IPC message receivers.
* Source/WebKit/UIProcess/Extensions/WebExtensionContext.messages.in:
  Add IPC messages SidebarSetActionClickBehavior and
SidebarGetActionClickBehavior.
* Source/WebKit/WebKit.xcodeproj/project.pbxproj: Add
  WebExtensionActionClickBehavior.h and
WebExtensionActionClickBehavior.serialization.in.
* Source/WebKit/WebProcess/Extensions/API/Cocoa/WebExtensionAPISidePanelCocoa.mm:
(WebKit::parseActionClickBehavior): Add helper function to parse the
behavior argument to sidePanel.setPanelBehavior.
(WebKit::WebExtensionAPISidePanel::getPanelBehavior): Add implementation.
(WebKit::WebExtensionAPISidePanel::setPanelBehavior): Add implementation.
* Tools/TestWebKitAPI/Tests/WebKitCocoa/WKWebExtensionAPISidebar.mm:
(TestWebKitAPI::TEST_F(WKWebExtensionAPISidebar, SidePanelSetPanelBehaviorPersists)):
Add test which ensures that the panel behavior configuration is
persisted.
(TestWebKitAPI::TEST_F(WKWebExtensionAPISidebar, SidePanelOpensSidebarOnActionClickWhenConfigured)):
Add test which ensures that the correct WebExtensionControllerDelegate
method is called based on the action icon&apos;s behavior configuration.

Canonical link: <a href="https://commits.webkit.org/283997@main">https://commits.webkit.org/283997@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/ccc999288b35e17742573973ead29d11d7071a3b

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/68049 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/47441 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/20708 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/72110 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/19194 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/70166 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/55237 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/19010 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/54396 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 win-tests~~](https://ews-build.webkit.org/#/builders/60/builds/12801 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/71116 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/43440 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/58806 "Passed tests") | [  ~~🧪 api-wpe~~](https://ews-build.webkit.org/#/builders/41/builds/34858 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🧪 webkitpy](https://ews-build.webkit.org/#/builders/6/builds/67560 "Passed tests") | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/40109 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/16222 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/17551 "Built successfully") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/62083 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/16563 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/73808 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/87/builds/12020 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/62/builds/15842 "Passed tests") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/61848 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/86/builds/12058 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/58881 "Passed tests") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/21/builds/61864 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/15116 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/88/builds/9763 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/119/builds/3376 "Passed tests") | | 
| | [  ~~🛠 tv~~](https://ews-build.webkit.org/#/builders/44/builds/43242 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
| | [  ~~🛠 tv-sim~~](https://ews-build.webkit.org/#/builders/46/builds/44315 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/45510 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/44057 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

xeenon commented on Source/WebKit/WebProcess/Extensions/API/Cocoa/WebExtensionAPISidePanelCocoa.mm:
```diff
@@ -73,6 +75,21 @@ static ParseResult parseWindowIdentifier(NSDictionary *options)
+{
+    static NSDictionary<NSString *, id> *types = @{
+        actionClickBehaviorKey: @YES.class,
+    };
```
New line after.

xeenon commented on Tools/TestWebKitAPI/Tests/WebKitCocoa/WKWebExtensionAPISidebar.mm:
```diff
@@ -1290,6 +1295,72 @@
+TEST_F(WKWebExtensionAPISidebar, SidePanelSetPanelBehaviorPersists)
+{
+    auto *script = @[
+        @"let panelBehavior = await browser.sidePanel.getPanelBehavior()",
```
Add some empty lines to space out the script for easier reading.

xeenon commented on Tools/TestWebKitAPI/Tests/WebKitCocoa/WKWebExtensionAPISidebar.mm:
```diff
@@ -1290,6 +1295,72 @@
+TEST_F(WKWebExtensionAPISidebar, SidePanelOpensSidebarOnActionClickWhenConfigured)
+{
+    auto *script = @[
+        @"await browser.sidePanel.setPanelBehavior({ openPanelOnActionClick: true })",
```
Ditto on empty lines.

AriYoung00 committed:
Implement sidePanel setPanelBehavior/getPanelBehavior API methods.
https://webkit.org/b/280021
rdar://135894275

Reviewed by Timothy Hatcher.

This patch implements the sidePanel.setPanelBehavior and
sidePanel.getPanelBehavior API methods. Additionally, it adds two tests
to verify the functionality of these APIs.

* Source/WebKit/DerivedSources-input.xcfilelist: Add
  WebExtensionActionClickbehavior.serialization.in.
* Source/WebKit/DerivedSources.make: Add
  WebExtensionActionClickBehavior.serialization.in.
* Source/WebKit/Scripts/webkit/messages.py:
(types_that_cannot_be_forward_declared): Add
WebKit::WebExtensionActionClickBehavior to
types_that_cannot_be_forward_declared.
* Source/WebKit/Shared/Extensions/WebExtensionActionClickBehavior.h: Added.
* Source/WebKit/Shared/Extensions/WebExtensionActionClickBehavior.serialization.in: Added.
* Source/WebKit/UIProcess/Extensions/Cocoa/API/WebExtensionContextAPISidebarCocoa.mm:
(WebKit::WebExtensionContext::sidebarSetActionClickBehavior): Add
IPC message receiver.
(WebKit::WebExtensionContext::sidebarGetActionClickBehavior): Add IPC
message receiver.
* Source/WebKit/UIProcess/Extensions/Cocoa/WebExtensionContextCocoa.mm:
(WebKit::WebExtensionContext::performAction): Add case to open the
sidebar if configured and it is possible.
* Source/WebKit/UIProcess/Extensions/WebExtensionContext.h: Add
  prototypes for new IPC message receivers.
* Source/WebKit/UIProcess/Extensions/WebExtensionContext.messages.in:
  Add IPC messages SidebarSetActionClickBehavior and
SidebarGetActionClickBehavior.
* Source/WebKit/WebKit.xcodeproj/project.pbxproj: Add
  WebExtensionActionClickBehavior.h and
WebExtensionActionClickBehavior.serialization.in.
* Source/WebKit/WebProcess/Extensions/API/Cocoa/WebExtensionAPISidePanelCocoa.mm:
(WebKit::parseActionClickBehavior): Add helper function to parse the
behavior argument to sidePanel.setPanelBehavior.
(WebKit::WebExtensionAPISidePanel::getPanelBehavior): Add implementation.
(WebKit::WebExtensionAPISidePanel::setPanelBehavior): Add implementation.
* Tools/TestWebKitAPI/Tests/WebKitCocoa/WKWebExtensionAPISidebar.mm:
(TestWebKitAPI::TEST_F(WKWebExtensionAPISidebar, SidePanelSetPanelBehaviorPersists)):
Add test which ensures that the panel behavior configuration is
persisted.
(TestWebKitAPI::TEST_F(WKWebExtensionAPISidebar, SidePanelOpensSidebarOnActionClickWhenConfigured)):
Add test which ensures that the correct WebExtensionControllerDelegate
method is called based on the action icon's behavior configuration.

Canonical link: https://commits.webkit.org/283997@main
